lib/Sema/CSSolver.cpp

```diff
@@ -1683,6 +1683,69 @@ getOperatorDesignatedNominalTypes(Constraint *bindOverload) {
   return operatorDecl->getDesignatedNominalTypes();
 }
 
+void ConstraintSystem::partitionForDesignatedTypes(
+    ArrayRef<Constraint *> Choices, ConstraintMatchLoop forEachChoice,
+    PartitionAppendCallback appendPartition,
+    SmallVectorImpl<SmallVector<unsigned, 4>> &definedInDesignatedType,
+    SmallVectorImpl<SmallVector<unsigned, 4>>
+        &definedInExtensionOfDesignatedType) {
+
+  auto designatedNominalTypes = getOperatorDesignatedNominalTypes(Choices[0]);
+  if (designatedNominalTypes.empty())
+    return;
+
+  auto examineConstraint =
+    [&](unsigned constraintIndex, Constraint *constraint) -> bool {
+    auto *decl = constraint->getOverloadChoice().getDecl();
+    auto *funcDecl = cast<FuncDecl>(decl);
+
+    auto *parentDecl = funcDecl->getParent()->getAsDecl();
+    for (auto designatedTypeIndex : indices(designatedNominalTypes)) {
+      auto *designatedNominal =
+        designatedNominalTypes[designatedTypeIndex];
+
+      if (parentDecl == designatedNominal) {
+        auto &constraints = definedInDesignatedType[designatedTypeIndex];
```

Comment:
I _think_ bodies of both of these `if`s could be unified into:

```c++
// preamble
auto *parentDC = functionDecl->getParent();
auto *parentDecl = parentDC->getSelfNominalTypeDecl();

// in the body
if (parentDecl == designatedNominal) {
  auto &container = parentDC->isExtensionContext() ? definedInExtensionOfDesignatedType : definedInDesignatedType;
  container[designatedTypeIndex].push_back(constraint);
  return true;
}

Reply:
Yes, something like this should work.